draft-barnes-acme.md

```diff
@@ -1212,66 +1212,74 @@ If all of the above verifications succeed, then the validation is successful.  I
 
-The ACME server's Challenge provides its random value R, and a random nonce used to identify the transaction:
+The ACME server's Challenge provides the following parameters:
 
 type (required, string):
 : The string "dvsni"
 
 r (required, string):
-: A random 32-byte octet, Base64-encoded
+: A random 32-byte octet, hex-encoded
 
 nonce (required, string):
 : A random 16-byte octet string, hex-encoded (so that it can be used as a DNS label)
 
+ports (required, array):
+: Integers representing acceptable ports to perform the challenge.
+
 ~~~~~~~~~~
 
 {
   "type": "dvsni",
-  "r": "Tyq0La3slT7tqQ0wlOiXnCY2vyez7Zo5blgPJ1xt5xI",
-  "nonce": "a82d5ff8ef740d12881f6d3c2277ab2e"
+  "r": "21ee2be8d2c201a08447e3d615f84fd5a1264222ecaefd183d1a8162150e0789",
+  "nonce": "a82d5ff8ef740d12881f6d3c2277ab2e",
+  "ports": [443],
 }
 
 ~~~~~~~~~~
 
-The client responds to this Challenge by configuring a TLS server on port 443 of a server with the domain name being validated:
+The client responds to this Challenge by configuring a TLS server on the chosen port of a server with the domain name being validated:
+
+1. Generate a self-signed certificate with a subjectAltName extension containing the dNSName value "\<R\>[0:32].\<R\>[32:64].acme.invalid"
+2. Compute a nonce domain name by appending the suffix ".acme.invalid" to the nonce provided by the server.
+3. Configure the TLS server such that when a client presents the nonce domain name in the SNI field, the server presents the generated certificate.
 
-1. Decode the server's random value R
-2. Generate a random 32-byte octet string S
-3. Compute Z = SHA-256(R &#124;&#124; S) (where &#124;&#124; denotes concatenation of octet strings)
-4. Generate a self-signed certificate with a subjectAltName extension containing two dNSName values:
-  1. The domain name being validated
-  2. A name formed by hex-encoding Z and appending the suffix ".acme.invalid"
-5. Compute a nonce domain name by appending the suffix ".acme.invalid" to the nonce provided by the server.
-6. Configure the TLS server such that when a client presents the nonce domain name in the SNI field, the server presents the generated certificate.
+The client SHOULD introduce its own entropy into each certificates (e.g. include a random serial number).
 
-The client's response provides its random value S:
+The client's responds with the desired port and key used within the certificate.:
 
 type (required, string):
 : The string "dvsni"
 
-s (required, string):
-: A random 32-byte secret octet string, Base64-encoded
+port (required, int):
+: Desired port to perform the challenge
```

Comment:
Why are we allowing ports != 443? That seems fairly hard to analyze from a security perspective. I notice that you don't even require that port < 1024.